### ISSUE TITLE
C++ std::vector<T> and Rust std::vec::Vec<T> support

### ISSUE THREAD

**Issue by myronahn** [dtolnay/cxx#33]
I've hacked in some preliminary `std::vector<T>` support to allow passing C++ vectors to Rust.

https://github.com/myronahn/cxx/tree/master

Right now it works for `std::vector<uint8_t>` since that was my primary need but it would be very easy to add in support for all the basic types.

Adding support for general types is a bit harder as the orphan rule is keeping me from defining macro-expanded trait impls for both the `Vector` class and the `UniquePtr` class.  I'm currently wondering what the best workaround would be for this.

It would be great to get a once-over from you if you have any spare time to take a look.  Is it worth continuing work on this, or are you in the middle of implementing something bigger/better?

Thanks in advance.

**Comment by dtolnay** [comment]
Wow this is terrific! I wasn't already working on this -- at the moment I am working on a test suite so that we have more extensive coverage of the various permutations of signatures: passing reference to an opaque Rust type to C++, passing a reference to an opaque C++ type to C++, reference to an opaque Rust type to Rust, reference to an opaque C++ type to Rust, *returning* a reference to an opaque type, passing a Box, returning a Box, passing a UniquePtr, returning a UniquePtr, etc, etc.

I am a bit hesitant to accept std::vector support in a form that can't be used for structs, so we would need to dig into those orphan rule issues. Do you have a commit that reproduces what you tried and how it failed?

**Comment by myronahn** [comment]
The test suite is definitely much-needed - I was running into all these various cases when I was working on the code!

Regarding the orphan rule, I think I might have been mistaken.  

I was trying to be clever and auto-generate the `Vector` trait impls for built-in types like `u8` using the macro expansion which obviously won't work in user code.  However it should work fine for user-defined structs/enums.  

I should be able to just generate the trait impls for built-in types inside the cxx crate itself and then let the macro expansion handle user-defined types in user code.  Let me give it another try to see if I can get it working.

**Comment by myronahn** [comment]
BTW thanks for writing this crate and for the quick response!

**Comment by myronahn** [comment]
So I've done a bit more work on this.

One of the big problems I had: I've been returning C++ `std::vector<T>`s to Rust wrapped in `std::unique_ptr<T>`s.  The issue is that I need to implement the `UniquePtrTarget` trait for `cxx::Vector<UserType>`.  

Both `UniquePtrTarget` and `cxx::Vector<T>` are in the `cxx` crate so the orphan rules say that the user can't implement `UniquePtrTarget` for `cxx::Vector<UserType>` (via macros) even though `UserType` is in the user crate.

I've solved this for now by creating a `RealVector<T>` that has the core implementation and a newtype `Vector<T>` which wraps `RealVector<T>` and is defined in the user crate via macros.  `UniquePtrTarget` is defined on `cxx::Vector<T>`

It's not ideal but it works.  Right now I have both a `cxx::Vector<T>` for passing C++ `std::vector<T>`s to Rust  and a `rust::RustVec<T>` (probably should rename that to `rust::Vec<T>`) for passing Rust `Vec<T>`s to C++.

I still need to manually define implementations for `cxx::Vector<T>`the built-in types (only done for `u8` now) but that's straightforward.  Would be good to modify the macro system to generate these in the `cxx` crate.

Would love ideas/feedback if you have time!

**Comment by dtolnay** [comment]
Regarding having to emit local wrappers, I haven't looked through all the code on your branch but I would have expected to see something like the following, which I didn't find.

```rust
// cxx
unsafe impl<T> UniquePtrTarget for CxxVector<T>
where
    T: VectorElement;

// generated code
unsafe impl VectorElement for #ty;
```

**Comment by myronahn** [comment]
I had been letting the `expand::expand_unique_ptr()` function takes care of generating the `UniquePtrTarget` implementation.  I'd assumed that I had to do this to match up with the C++ functions generated by `write::write_unique_ptr()` for a particular `std::vector<T>`.  

So for every `T` there will be a different `impl UniquePtrTarget` for `cxx::Vector<T>` which correspond to a set of C++ functions generated by `write::write_unique_ptr()` for `std::vector<T>`.

If there's a way to do it generically, that would be great.

**Comment by myronahn** [comment]
BTW I've also added some tests to your test framework which demonstrate the `UniquePtr<Vector<T>>` as function parameter and return value.

**Comment by myronahn** [comment]
I should clarify: it would be great to have a generic implementation of `UniquePtrTarget` for `cxx::Vector<T>` but I'm not sure it is possible given the fact that the implementation needs to call non-template C++ functions.

**Comment by myronahn** [comment]
I've created a pull request for easier review with no expectations of it being accepted.

**Comment by myronahn** [comment]
I had an idea to eliminate the need for a newtype: if you define a `GenericTarget` which would handle the needed methods for all possible combinations of containers (`Vector<T>`, `UniquePtr<T>`, `UniquePtr<Vector<T>>, etc `where T: GenericTarget)  then you might be able to avoid the newtype.  Seems like it could explode though.  We might be able to get away with only defining the ones that are used.

**Comment by dtolnay** [comment]
Basic support implemented in #67 + #148. There is a lot of work still left for a good binding (for example `rust::Vec<T>::operator[]` is pretty necessary) but I'll go ahead and close this and we can follow up in more specific issues/PRs.
